### ISSUE TITLE
Consider having expression interpreter available in all build flavors

### ISSUE THREAD

**Issue by bartdesmet** [dotnet/corefx#3244]
For some of our Bing service, we often run into the case where expression compilation is too expense, either because it's on a request/response path, or because it's statically known that the expression is small and will be evaluated only once. I can provide more details on such scenarios if needed.

Unfortunately, expression interpretation is a property of the .NET build flavor, e.g. for .NET Native the Compile method uses the interpreter.

It'd be nice to have an Interpret method alongside Compile to enable those scenarios. It'd also help a lot in writing parity tests for the expression compiler and interpreter because their outcomes can be cross-checked.

**Comment by bartdesmet** [comment]
Discussed offline with @VSadov; broadening the audience here now.

**Comment by VSadov** [comment]
@jaredpar @madstorgersen - FYI.

**Comment by VSadov** [comment]
I think allowing compiling and interpreting behavior to be both possible in the same binary would be useful. Currently these are mutually exclusive - I.E. you can have System.Linq.Expressions that either has compiler or interpreter. 

There could be some technical difficulties in implementation, but hopefully these can be resolved.

We would need some new API to get to the interpreter. Most obvious would be ".Interpret" alongside with Compile, but it also could be an overload of Compile like Compile(bool preferInterpretation) or something like that.

In terms of testing - We need to test both compiler and interpreter anyways and this change will only make that easier.

**Comment by antiufo** [comment]
I recently noticed that Expression.Compile() was in the hot path of my software project (many small, single-use compilations). I refactored out the interpreter and made it available to Net45/Dnxcore50 (nuget: Shaman.System.Linq.Expressions.Interpreter). The whole thing now runs much faster.
I agree with the idea of making the interpreter publically available.

**Comment by VSadov** [comment]
We have discussed this and it seems a good thing to have.
As for the entry point, the preferred shape would be "Compile(bool preferInterpretation)"

The reasons are:
1) Compile() does not necessary compile. It basically prefers compiling if available, but otherwise will interpret.
2) The method does not really start interpretation. Technically it still compiles into a delegate, just with a different form that will be executed by the means of interpretation.

Compile(bool preferInterpretation) should use interpreter when preferInterpretation == true and interpreter is available (I see no reasons for it to be unavailable, but who knows..)

When preferInterpretation == false, it behaves like the old Compile() - it will use compiler if available and interpreter otherwise.

**Comment by VSadov** [comment]
@terrajobst @stephentoub @jaredpar  - this will need to be officially reviewed. Addition of public API.

**Comment by VSadov** [comment]
CC @madstorgersen @billchi-ms @dinov

**Comment by DinoV** [comment]
This could also just be an extension method that the interpreter DLL makes available instead of being a new method on LambdaExpression.  Then users can add a reference.

**Comment by bartdesmet** [comment]
An extension method is a design option indeed, and it'd keep the assembly size smaller, but there are a few hurdles to overcome if we decide to go that way:

- Every build flavor would need to have the method as an extension method in order for binaries compiled against it to be binary compatible / portable across multiple framework build flavors.
- The assembly would need InternalsVisibleTo access to the System.Linq.Expressions assembly because there's a few calls to internal members, e.g. on BinaryExpression to reduce.
- For the Compile() instance method on build flavors that don't support compilation, we'd need a reference the opposite way (cyclic) or include the code in System.Linq.Expressions as well.

**Comment by terrajobst** [comment]
We just reviewed this and it looks good as proposed.
